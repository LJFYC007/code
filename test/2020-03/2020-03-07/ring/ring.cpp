#include<bits/stdc++.h>

using namespace std;

#define REP(i, a, b) for ( int i = (a), _end_ = (b); i <= _end_; ++ i )
#define mem(a) memset((a), 0, sizeof(a))
#define str(a) strlen(a)
#define pii pair<int, int>
#define int long long
typedef long long LL;

const int maxn = 2e5 + 10;

int n, m, a[maxn], f[maxn], g[maxn];

inline int get(int l, int r)
{
	int sum = g[r] - g[l + 1];
	return sum + 2 * a[l] * a[l + 1];
}

signed main()
{
	freopen("ring.in", "r", stdin);
	freopen("ring.out", "w", stdout);
	scanf("%lld%lld", &n, &m);
	REP(i, 1, n) { scanf("%lld", &a[i]); f[0] += a[i] * a[i]; }
	sort(a + 1, a + n + 1);
	REP(i, 3, n) g[i] = g[i - 1] + a[i] * (a[i - 1] + a[i - 2]) - a[i - 1] * a[i - 2];
	REP(i, 1, n / 2)
	{
		if ( n % i != 0 ) continue ;
		int x = i, l = 1, r = n / x;
		REP(j, 1, x)
		{
			f[i] += get(l, r);
			l += n / x; r += n / x;
		}
	}
	REP(i, 1, m)
	{
		int k; scanf("%lld", &k);
		if ( k == 0 ) { printf("%lld\n", f[k]); continue ; }
		printf("%lld\n", f[__gcd(k, n)]);
	}
	return 0;
}
