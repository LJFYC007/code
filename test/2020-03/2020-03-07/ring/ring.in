100 51
42 68 35 1 70 25 79 59 63 65 6 46 82 28 62 92 96 43 28 37 92 5 3 54 93 83 22 17 19 96 48 27 72 39 70 13 68 100 36 95 4 12 23 34 74 65 42 12 54 69 48 45 63 58 38 60 24 42 30 79 17 36 91 43 89 7 41 43 65 49 47 6 91 30 71 51 7 2 94 49 30 24 85 55 57 41 67 77 32 9 45 40 27 24 38 39 19 83 30 42 
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
