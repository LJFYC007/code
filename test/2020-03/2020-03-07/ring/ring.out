296800
296516
296516
296516
296517
296535
296516
296516
296517
296516
296566
296516
296517
296516
296516
296535
296517
296516
296516
296516
296593
296516
296516
296516
296517
296582
296516
296516
296517
296516
296566
296516
296517
296516
296516
296535
296517
296516
296516
296516
296593
296516
296516
296516
296517
296535
296516
296516
296517
296516
296668
